File: leetcode/003.py
#   003
#   Longest Substring Without Repeating Characters
#####################################################
class Solution:
    # @return an integer
    def lengthOfLongestSubstring(self, s):
        longest = 0
        tmp = 0
        head = 0
        tail = 0
        length = len(s)
        while tail < length:
            c = s[tail]
            index = s.find(c, head, tail)
            if index == -1:
                tmp += 1
            elif index == head:
                head += 1
            else:
                if tmp > longest:
                    longest = tmp
                head = index + 1
                tmp = tail - head + 1
            tail += 1
            print head, tail, s[head:tail]
        if tmp > longest:
            longest = tmp
        return longest

#####################################################

s = "epqcrkeyjzbkvkrganbg"
print s

sol = Solution()
print sol.lengthOfLongestSubstring(s)

File: leetcode/003.rb
# 003
# Longest Substring Without Repeating Characters
#####################################################
# @param {string} s
# @return {Integer}
def length_of_longest_substring(s)
  longest = 0
  tmp = 0
  head = 0
  tail = 0

  while tail < s.length
    c = s[tail]
    if not s[head...tail].include? c
      tmp += 1
    elsif c == s[head]
      head += 1
    else
      longest = tmp if tmp > longest
      head += s[head...tail].index(c) + 1
      tmp = tail - head + 1
    end
    puts "#{head} #{tail} #{tmp}"
    tail += 1
  end
  longest = tmp if tmp > longest
  longest
end

#####################################################

s = "epqcrkeyjzbkvkrganbg"
puts s
puts length_of_longest_substring(s)

